Add Azure Red Hat OpenShift 2024-08-12-preview API version

cadenmarchese opened the pull request:
# ARM (Control Plane) API Specification Update Pull Request 

> [!TIP]
> Overwhelmed by all this guidance? See the `Getting help` section at the bottom of this PR description.

## PR review workflow diagram

Please understand this diagram before proceeding. It explains how to get your PR approved & merged.

![spec_pr_review_workflow_diagram](https://github.com/Azure/azure-rest-api-specs/assets/4429827/5bb5e7ce-8aff-4dbb-a3f8-0d9b68fef5b1)

## Purpose of this PR

What's the purpose of this PR? Check the specific option that applies. This is **mandatory**!

  - [ ] New resource provider.
  - [x] New API version for an existing resource provider. (If API spec is not defined in TypeSpec, the PR should have been created in adherence to [OpenAPI specs PR creation guidance](https://aka.ms/azsdkdocs/createopenapispec)).
  - [ ] Update existing version for a new feature. (This is applicable only when you are revising a private preview API version.)
  - [ ] Update existing version to fix OpenAPI spec quality issues in S360.
  - [ ] Convert existing [OpenAPI spec to TypeSpec spec](https://aka.ms/typespec/conversion) (do not combine this with implementing changes for a new API version).
  - [ ] Other, please clarify:
    - _edit this with your clarification_

## Due diligence checklist

To merge this PR, you **must** go through the following checklist and confirm you understood 
and followed the instructions by checking all the boxes:

- [x] I confirm this PR is modifying Azure Resource Manager (ARM) related specifications, and not data plane related specifications.
- [x] I have reviewed following [Resource Provider guidelines](https://aka.ms/rpguidelines), including
  [ARM resource provider contract](https://aka.ms/azurerpc) and
  [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md) (estimated time: 4 hours).  
  I understand this is required before I can proceed to the diagram Step 2, "ARM API changes review", for this PR.

## Additional information

<details>
<summary> Viewing API changes</summary>

For convenient view of the API changes made by this PR, refer to the URLs provided in the table 
in the `Generated ApiView` comment added to this PR. You can use ApiView to show API versions diff. 

</details>
<details>
<summary>Suppressing failures</summary>

If one or multiple validation error/warning suppression(s) is detected in your PR, please follow the 
[suppressions guide](https://aka.ms/azsdk/pr-suppressions) to get approval.

</details>

## Getting help

- First, please carefully read through this PR description, from top to bottom. Please fill out the `Purpose of this PR` and `Due diligence checklist`.
- If you don't have permissions to remove or add labels to the PR, request `write access` per [aka.ms/azsdk/access#request-access-to-rest-api-or-sdk-repositories](https://aka.ms/azsdk/access#request-access-to-rest-api-or-sdk-repositories)
- To understand what you must do next to merge this PR, see the `Next Steps to Merge` comment. It will appear within few minutes of submitting this PR and will continue to be up-to-date with current PR state.
- For guidance on fixing this PR CI check failures, see the hyperlinks provided in given failure 
  and https://aka.ms/ci-fix.
- For help with ARM review (PR workflow diagram Step 2), see https://aka.ms/azsdk/pr-arm-review.
- If the PR CI checks appear to be stuck in `queued` state, please add a comment with contents `/azp run`.
  This should result in a new comment denoting a `PR validation pipeline` has started and the checks should be updated after few minutes.
- If the help provided by the previous points is not enough, post to https://aka.ms/azsdk/support/specreview-channel and link to this PR.

cadenmarchese committed:
commit latest API version first
* ensures reviewers can see changes from new API

cadenmarchese committed:
commit 2024-08-12-preview changes

commit changes to readme

* this file appears required for comparison with Azure types

commit v2024-08-12-preview examples

add descriptions to all new api fields

adjust one of the code comments according to feedback on the Red Hat side

add platformworkloadidentityroleset + examples

fix relative path to common-types
use common-types files directly rather than copying them into
the redhatopenshift directory

use common-types/v6, use ManagedServiceIdentity type definition

fix Swagger SpellCheck

remove "format":"uuid" from clientIds, principalIds, tenantIds
attempt to fix Swagger LintDiff check

convert platformWorkloadIdentities to map

identity -> managedServiceIdentity

rename managedServiceIdentity back to identity, use common-types
directly

fix UNRESOLVABLE_REFERENCE

fix relative path to common types

attempt to fix UNRESOLVABLE_REFERENCE

fix OBJECT_MISSING_REQUIRED_PROPERTY by fixing examples

attempt to fix PatchBodyParametersSchema

ms-bogdan-rotaru commented on specification/redhatopenshift/resource-manager/Microsoft.RedHatOpenShift/openshiftclusters/preview/2024-08-12-preview/redhatopenshift.json:
```diff
@@ -0,0 +1,2722 @@
+        },
+        "principalId": {
+          "type": "string"
+        }
```
Can you please add description for these fields?

ms-bogdan-rotaru commented on specification/redhatopenshift/resource-manager/Microsoft.RedHatOpenShift/openshiftclusters/preview/2024-08-12-preview/redhatopenshift.json:
```diff
@@ -0,0 +1,2722 @@
+          "additionalProperties": {
+            "$ref": "#/definitions/ClusterUserAssignedIdentity"
+          }
+        }
```
Can you please add description for these fields?

ms-bogdan-rotaru commented on specification/redhatopenshift/resource-manager/Microsoft.RedHatOpenShift/openshiftclusters/preview/2024-08-12-preview/redhatopenshift.json:
```diff
@@ -0,0 +1,2722 @@
+          "type": "object",
+          "additionalProperties": {
+            "$ref": "#/definitions/ClusterUserAssignedIdentity"
+          }
```
Use of `additionalProperties` is not allowed for properties owned by the service. The only time its ok to use it is when the properties are pass thru (user defined) and not subject to any validations. Please make this is an array or provide an explanation for why you need this.

cadenmarchese replied:
Hi @ms-bogdan-rotaru, thanks for the review. in this case and according to my understanding, the array model is deprecated by the managed identity team: https://eng.ms/docs/products/arm/rbac/managed_identities/msionboardinguserassigned#api-requirements--design that is why we are using a dictionary/map instead which is probably what is resulting in our swagger generation adding the ClusterUserAssignedIdentity as `additionalProperties`. Our API was designed with the MSI/managed identity team's requirements in mind.

If you take a look at our API definition, you can see how we've implemented this requirement: https://github.com/Azure/ARO-RP/pull/3727/files?diff=unified&w=0#diff-c7bfe4dfde8a1a575263edc226847cbab8ac7f60f04193f531a17d73571a3398R438-R458 and how AKS RP has done the same for a similar use case: https://msazure.visualstudio.com/CloudNativeCompute/_git/aks-rp?path=/resourceprovider/server/microsoft.com/containerservice/datamodel/v20240602preview/types.go&version=GBmaster&line=217&lineEnd=247&lineStartColumn=1&lineEndColumn=1&lineStyle=plain&_a=contents

That being said, I'm not totally clear on what you mean by pass thru properties or validations. Can you explain these points more?

ms-bogdan-rotaru commented on specification/redhatopenshift/resource-manager/Microsoft.RedHatOpenShift/openshiftclusters/preview/2024-08-12-preview/redhatopenshift.json:
```diff
@@ -0,0 +1,2722 @@
+        "objectId": {
+          "type": "string",
+          "readOnly": true
+        }
```
Can you please add description for these fields?

cadenmarchese commented on specification/redhatopenshift/resource-manager/Microsoft.RedHatOpenShift/openshiftclusters/preview/2024-08-12-preview/redhatopenshift.json:
```diff
@@ -0,0 +1,2722 @@
+        },
+        "principalId": {
+          "type": "string"
+        }
```
They are GUID values and their type will be enforced in the API itself via validation. Here's an example of how we do it: https://github.com/Azure/ARO-RP/blob/master/pkg/api/v20240812preview/openshiftcluster_validatestatic.go#L452-L482

So we would have a similar function in this file that validates identity values are GUID before writing the data to cosmos, but we haven't implemented it yet.

cadenmarchese commented:
/azp run

cadenmarchese commented:
This PR had a previous ARM sign-off, but I added a few new commits, so I'm re-adding the `WaitForARMFeedback` label here - please take a look and let me know if anything needs to be adjusted. I believe that the Swagger SpellCheck can be disregarded.

gary-x-li commented:
>     "PlatformWorkloadIdentityRoleSetUpdate": {

It almost the same as `PlatformWorkloadIdentityRoleSetUpdate` (except it doesn't have the `allOf`) and it doesn't seem to be used.

---
Refers to: specification/redhatopenshift/resource-manager/Microsoft.RedHatOpenShift/openshiftclusters/preview/2024-08-12-preview/redhatopenshift.json:2507 in d2f1c82. [](commit_id = d2f1c82be0d77127469ba5d064b283e9c2a93ad1, deletion_comment = False)

cadenmarchese commented:
@gary-x-li where are you seeing the extra/unused `PlatformWorkloadIdentityRoleSetUpdate`? I only see `PlatformWorkloadIdentityRoleSetUpdate` appear once in the generated swagger file (specification/redhatopenshift/resource-manager/Microsoft.RedHatOpenShift/openshiftclusters/preview/2024-08-12-preview/redhatopenshift.json) and only in the latest commit [d2f1c82](https://github.com/Azure/azure-rest-api-specs/pull/30004/commits/d2f1c82be0d77127469ba5d064b283e9c2a93ad1) (not in the referenced commit https://github.com/Azure/azure-rest-api-specs/commit/7838b6a246f5be0970df7f5c1283ecbd0022a002)

razvanbadea-msft commented on specification/redhatopenshift/common-types/resource-management/v3/types.json:
```diff
@@ -0,0 +1,695 @@
+  "swagger": "2.0",
+  "info": {
+    "version": "3.0",
+    "title": "Common types"
```
>title": "Common types" [](http://example.com/codeflow?start=5&length=22)

Why are you not using the common types file directly?

cadenmarchese replied:
I fixed the issue where we weren't using the common-types files directly, but the migration to the latest common-types files `v6` are resulting in breaking change (cross version) CI errors:
```
The new version has a different format 'uuid' than the previous one ''.
New: [common-types/resource-management/v6/types.json#L591:5](https://github.com/Azure/azure-rest-api-specs/blob/6819c7a7f114669cefeb12959712d03ffd29a287/specification/common-types/resource-management/v6/types.json#L591:5)
parameters.SubscriptionIdParameter
Old: [common-types/resource-management/v3/types.json#L622:5](https://github.com/Azure/azure-rest-api-specs/blob/main/specification/common-types/resource-management/v3/types.json#L622:5)
```
There are also similar failures as a result of our most recent API using the v3 types when this PR is now expecting the v6 types:
```
The new version has a different format 'uuid' than the previous one ''.
New: [openshiftclusters/preview/2024-08-12-preview/redhatopenshift.json#L1201:9](https://github.com/Azure/azure-rest-api-specs/blob/6819c7a7f114669cefeb12959712d03ffd29a287/specification/redhatopenshift/resource-manager/Microsoft.RedHatOpenShift/openshiftclusters/preview/2024-08-12-preview/redhatopenshift.json#L1201:9)
paths['/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.RedHatOpenShift/openshiftclusters/{resourceName}/secret/{childResourceName}'].patch.parameters
Old: [openshiftclusters/stable/2023-11-22/redhatopenshift.json#L1158:9](https://github.com/Azure/azure-rest-api-specs/blob/main/specification/redhatopenshift/resource-manager/Microsoft.RedHatOpenShift/openshiftclusters/stable/2023-11-22/redhatopenshift.json#L1158:9)
```
Is it safe to disregard the errors from CI or do I need to revert to v3? Is it required to use the latest common-types? I got a previous sign-off with the v3 types, so I assume they are not deprecated.

razvanbadea-msft replied:
it is not deprecated but strongly suggested to do it.

you can revert to v3 from common-types so that this PR will be merged faster and  keep in mind for updating to latest version for the next PR as there are some minor breaking changes because of it as i can see.

cadenmarchese replied:
Thanks. If we were to stick with v6 for the sake of not needing to do it later, would the breaking changes as a result be OK to merge, or is there some action needed to resolve them?

razvanbadea-msft replied:
You will have to get an approval from Azure Breaking Change Review Board. - https://aka.ms/brch

cadenmarchese replied:
@razvanbadea-msft I have one more question, and thanks for your help. Let's say we moved forward with v3 for now, and onboard `Identity` definition as-is. What is the path to migrate to v6 in a later API version given that there are new required definitions for MSI in v6 (ManagedServiceIdentity as you said)? Would we be able to keep the `Identity` definition as-is at that point to avoid breaking things on our end or would we need to migrate somehow?

razvanbadea-msft replied:
You can have a custom one but even in v3 you can use this type in this pr so that when you will upgrade to a new version of common types you can just change the version for it https://github.com/Azure/azure-rest-api-specs/blob/main/specification/common-types/resource-management/v3/managedidentity.json

razvanbadea-msft commented:
>             "$ref": "../../../../../common-types/resource-management/v3/types.json#/parameters/ApiVersionParameter"

you should use the latest version of common types

---
Refers to: specification/redhatopenshift/resource-manager/Microsoft.RedHatOpenShift/openshiftclusters/preview/2024-08-12-preview/redhatopenshift.json:29 in d2f1c82. [](commit_id = d2f1c82be0d77127469ba5d064b283e9c2a93ad1, deletion_comment = False)

razvanbadea-msft commented:
>     "Identity": {

Why are you not using the ManagedServiceIdentity from common-types?

---
Refers to: specification/redhatopenshift/resource-manager/Microsoft.RedHatOpenShift/openshiftclusters/preview/2024-08-12-preview/redhatopenshift.json:1947 in d2f1c82. [](commit_id = d2f1c82be0d77127469ba5d064b283e9c2a93ad1, deletion_comment = False)

razvanbadea-msft commented:
> > ```
> > "Identity": {
> > ```
> 
> Why are you not using the ManagedServiceIdentity from common-types?
> 
> Refers to: specification/redhatopenshift/resource-manager/Microsoft.RedHatOpenShift/openshiftclusters/preview/2024-08-12-preview/redhatopenshift.json:1947 in [d2f1c82](https://github.com/Azure/azure-rest-api-specs/commit/d2f1c82be0d77127469ba5d064b283e9c2a93ad1). [](commit_id = [d2f1c82](https://github.com/Azure/azure-rest-api-specs/commit/d2f1c82be0d77127469ba5d064b283e9c2a93ad1), deletion_comment = False)

I saw also that you are defining an Identity similar with the ManagedServiceIdentity one.

cadenmarchese commented:
@razvanbadea-msft after talking with my team I've reverted this PR to use common-types/v3 with the goal of migrating to v6 in a future PR. The justification for doing this is that moving to v6 involves unplanned breaking changes, and moving to the ManagedServiceIdentity definition involves significant code changes on our side that prevent this from being feasible right now given that we're on a tight timeline to deliver a feature.

razvanbadea-msft commented:
/azp run

razvanbadea-msft commented:
the required Swagger SpellCheck fails, please fix it first

cadenmarchese committed:
suppress avoidAdditionalProperties
* recieved signoff from ARM team to use additionalProperties
during API modeling call

cadenmarchese committed:
fix SDK generation (duplicate type definitions)

cadenmarchese committed:
fix SDK generation

cadenmarchese committed:
migrate all references to common-types v6
attempt to fix SDK generation

cadenmarchese committed:
fix ModelValidation

mikeharder commented on specification/redhatopenshift/resource-manager/Microsoft.RedHatOpenShift/openshiftclusters/preview/2024-08-12-preview/redhatopenshift.json:
```diff
@@ -0,0 +1,2732 @@
+            "description": "The OpenShift cluster resource.",
+            "required": true,
+            "schema": {
+              "$ref": "#/definitions/OpenShiftClusterUpdate"
```
Start of LintDiff error `PatchBodyParametersSchema`

mikeharder commented on specification/redhatopenshift/resource-manager/Microsoft.RedHatOpenShift/openshiftclusters/preview/2024-08-12-preview/redhatopenshift.json:
```diff
@@ -0,0 +1,2732 @@
+          "x-ms-client-flatten": true
+        },
+        "identity": {
+          "$ref": "../../../../../../common-types/resource-management/v6/managedidentity.json#/definitions/ManagedServiceIdentity",
```
LintDiff `PatchBodyParametersSchema`: Step 2

Contains a required property `type` on L70:

https://github.com/Azure/azure-rest-api-specs/blob/8af8cd76b7349f357b8ea4f948ce2ee2aac318bb/specification/common-types/resource-management/v6/managedidentity.json#L41-L72

cadenmarchese replied:
Thanks @mikeharder. Should we not be referencing the common-types file for OpenShiftClusterUpdate?

I ask because it seems like if we pull it out, we will need to redefine ManagedServiceIdentity in our swagger file, which will result in duplicate schema errors in SDK generation.

mikeharder replied:
I don't know.  You can try searching the repo for specs similar to yours, to see what they do.  Or, you can ask at the ARM Office Hours: https://aka.ms/armofficehoursinfo

cadenmarchese replied:
Thanks - after looking at how other RPs do this, I've suppressed the failure with the same justification provided by them. Our swagger spec for `Update` is the same as other RPs, and, we get the patch body from common-types and the type is indeed patchable, so this appears to be a false positive.

cadenmarchese committed:
suppress PatchBodyParametersSchema as this patch body comes from
common-types and is required

ramoka178 commented:
/azp run

cadenmarchese commented:
@mikekistler do you mind taking another look at the breaking changes? They are all still related to our migration from common-types `v3` to common-types `v6`, with the exception of the removal of `systemData` where it was not required and included in previous versions in error.

mikekistler commented:
@cadenmarchese Pls come to Breaking Change office hours next Monday at 10AM PT to discuss these changes. I have concerns with the RemovedProperty and AddedRequiredProperty issues flagged in the Breaking Changes report.

cadenmarchese committed:
correct platformworkloadidentityrolesets to be plural

azure-sdk commented:
**API change check**

APIView has identified API level changes in this PR and created following API reviews.

[Microsoft.RedHatOpenShift-openshiftclusters](https://apiview.dev/Assemblies/Review/73b89a72e9bf4c968777c5d0e891ab37?revisionId=b11beae61e4443af9722ef240aac6a8e)

mentat9 commented on specification/redhatopenshift/resource-manager/Microsoft.RedHatOpenShift/openshiftclusters/readme.md:
```diff
@@ -26,34 +26,12 @@ These are the global settings for the Azure Red Hat OpenShift API.
-  - preview/2023-07-01-preview/redhatopenshift.json
+tag: package-2024-08-12-preview
+suppressions:
+  - code: AvoidAdditionalProperties
```
>AvoidAdditionalProperties [](http://example.com/codeflow?start=10&length=25)

[ARMBlockingComment] Please add a where: clause. Also, please give the actual reason this was allowed by the ARM reviewer(s). From comments it sounds like it's due to a "passthru" API. IOW, your API is passing parameters through (unvalidated) to an external API you don't control.

mentat9 commented on specification/redhatopenshift/resource-manager/Microsoft.RedHatOpenShift/openshiftclusters/readme.md:
```diff
@@ -26,34 +26,12 @@ These are the global settings for the Azure Red Hat OpenShift API.
+suppressions:
+  - code: AvoidAdditionalProperties
+    reason: Our use of additionalProperties is valid according to review from ARM team
+  - code: PatchBodyParametersSchema
```
>PatchBodyParametersSchema [](http://example.com/codeflow?start=10&length=25)

[ARMBlockingComment] Add where: clause.

mentat9 commented on specification/redhatopenshift/resource-manager/Microsoft.RedHatOpenShift/openshiftclusters/preview/2024-08-12-preview/redhatopenshift.json:
```diff
@@ -0,0 +1,2732 @@
+        }
+      }
+    },
+    "/subscriptions/{subscriptionId}/providers/Microsoft.RedHatOpenShift/locations/{location}/openshiftversions": {
```
>openshiftversions [](http://example.com/codeflow?start=94&length=17)

[ARMBlockingComment] Resource type names should always use camelCase. Comment applies to all resource types defined in this file.

mentat9 commented on specification/redhatopenshift/resource-manager/Microsoft.RedHatOpenShift/openshiftclusters/preview/2024-08-12-preview/redhatopenshift.json:
```diff
@@ -0,0 +1,2732 @@
+        }
+      }
+    },
+    "/subscriptions/{subscriptionId}/providers/Microsoft.RedHatOpenShift/locations/{location}/openshiftversions": {
```
>/subscriptions/{subscriptionId}/providers/Microsoft.RedHatOpenShift/locations/{location}/openshiftversions [](http://example.com/codeflow?start=5&length=106)

[ARMBlockingComment] Need to add a corresponding point/singleton GET path as required by ARM/RPC: [https://github.com/cloud-and-ai-microsoft/resource-provider-contract/blob/master/v1.0/get-resource.md](https://github.com/cloud-and-ai-microsoft/resource-provider-contract/blob/master/v1.0/get-resource.md)

cadenmarchese replied:
Done, thanks!

mentat9 commented on specification/redhatopenshift/resource-manager/Microsoft.RedHatOpenShift/openshiftclusters/preview/2024-08-12-preview/redhatopenshift.json:
```diff
@@ -0,0 +1,2732 @@
+        }
+      }
+    },
+    "/subscriptions/{subscriptionId}/providers/Microsoft.RedHatOpenShift/locations/{location}/platformworkloadidentityrolesets": {
```
>/subscriptions/{subscriptionId}/providers/Microsoft.RedHatOpenShift/locations/{location}/platformworkloadidentityrolesets [](http://example.com/codeflow?start=5&length=121)

[ARMBlockingComment] Need to add a singleton/point GET.

cadenmarchese replied:
Done, thanks!

mentat9 commented on specification/redhatopenshift/resource-manager/Microsoft.RedHatOpenShift/openshiftclusters/preview/2024-08-12-preview/redhatopenshift.json:
```diff
@@ -0,0 +1,2732 @@
+        }
+      }
+    },
+    "/subscriptions/{subscriptionId}/providers/Microsoft.RedHatOpenShift/locations/{location}/platformworkloadidentityrolesets": {
```
>platformworkloadidentityrolesets [](http://example.com/codeflow?start=94&length=32)

[ARMBlockingComment] Must be camelCase (platformWorkloadIdentityRoleSets).

mentat9 commented on specification/redhatopenshift/resource-manager/Microsoft.RedHatOpenShift/openshiftclusters/preview/2024-08-12-preview/redhatopenshift.json:
```diff
@@ -0,0 +1,2732 @@
+        }
+      }
+    },
+    "/subscriptions/{subscriptionId}/providers/Microsoft.RedHatOpenShift/locations/{location}/platformworkloadidentityrolesets": {
```
>platformworkloadidentityrolesets [](http://example.com/codeflow?start=94&length=32)

How are these created and deleted when you don't have PUT or DELETE?

cadenmarchese replied:
This is a proxy resource that is only updated internally, not via our public-facing APIs.

mentat9 commented on specification/redhatopenshift/resource-manager/Microsoft.RedHatOpenShift/openshiftclusters/preview/2024-08-12-preview/redhatopenshift.json:
```diff
@@ -0,0 +1,2732 @@
+        }
+      }
+    },
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.RedHatOpenShift/openshiftclusters/{resourceName}/machinePool/{childResourceName}": {
```
>machinePool [](http://example.com/codeflow?start=141&length=11)

[ARMBlockingComment] Name must be plural and match the corresponding collection GET API.

cadenmarchese replied:
After a conversation with my team, we've decided to remove resource types machinePool, syncSet, secret, and syncIdentityProvider as they are unused. The endpoints sit behind an AFEC feature flag, our RP's API doesn't have any routes or handling for them, they are not part of our ARM manifests, and they were never available to customers.

mentat9 commented on specification/redhatopenshift/resource-manager/Microsoft.RedHatOpenShift/openshiftclusters/preview/2024-08-12-preview/redhatopenshift.json:
```diff
@@ -0,0 +1,2732 @@
+            "pattern": "^[a-zA-Z0-9]$|^[a-zA-Z0-9][-_a-zA-Z0-9]*[a-zA-Z0-9]$"
+          },
+          {
+            "name": "childResourceName",
```
>childResourceName [](http://example.com/codeflow?start=21&length=17)

[ARMBlockingComment] Name is too generic. Should be machinePoolName or similar.

cadenmarchese replied:
see comment https://github.com/Azure/azure-rest-api-specs/pull/30004#discussion_r1887283477

mentat9 commented on specification/redhatopenshift/resource-manager/Microsoft.RedHatOpenShift/openshiftclusters/preview/2024-08-12-preview/redhatopenshift.json:
```diff
@@ -0,0 +1,2732 @@
+        }
+      }
+    },
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.RedHatOpenShift/openshiftclusters/{resourceName}/secret/{childResourceName}": {
```
>secret [](http://example.com/codeflow?start=141&length=6)

Secrets are not allowed in GET payloads. Must use list* POST action.

cadenmarchese replied:
see comment https://github.com/Azure/azure-rest-api-specs/pull/30004#discussion_r1887283477

mentat9 commented on specification/redhatopenshift/resource-manager/Microsoft.RedHatOpenShift/openshiftclusters/preview/2024-08-12-preview/redhatopenshift.json:
```diff
@@ -0,0 +1,2732 @@
+        }
+      }
+    },
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.RedHatOpenShift/openShiftCluster/{resourceName}/secrets": {
```
>secrets [](http://example.com/codeflow?start=140&length=7)

[ARMBlockingComment] Secrets are not allowed in GET payloads. If you need an API to retrieve secrets, it needs to be a list<thing> POST action.

cadenmarchese replied:
see comment https://github.com/Azure/azure-rest-api-specs/pull/30004#discussion_r1887283477

mentat9 commented on specification/redhatopenshift/resource-manager/Microsoft.RedHatOpenShift/openshiftclusters/preview/2024-08-12-preview/redhatopenshift.json:
```diff
@@ -0,0 +1,2732 @@
+        }
+      }
+    },
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.RedHatOpenShift/openShiftCluster/{resourceName}/secrets": {
```
>openShiftCluster [](http://example.com/codeflow?start=108&length=16)

[ARMBlockingComment] openShiftClusters (with an 's')

mentat9 commented on specification/redhatopenshift/resource-manager/Microsoft.RedHatOpenShift/openshiftclusters/preview/2024-08-12-preview/redhatopenshift.json:
```diff
@@ -0,0 +1,2732 @@
+            "$ref": "../../../../../../common-types/resource-management/v6/types.json#/parameters/ResourceGroupNameParameter"
+          },
+          {
+            "name": "resourceName",
```
>resourceName [](http://example.com/codeflow?start=21&length=12)

[ARMBlockingComment] This name is too generic. Should be "openShiftClusterName" or similar. Needs to be fixed on all APIs in this file.

cadenmarchese replied:
see comment https://github.com/Azure/azure-rest-api-specs/pull/30004#discussion_r1887283477

mentat9 commented on specification/redhatopenshift/resource-manager/Microsoft.RedHatOpenShift/openshiftclusters/preview/2024-08-12-preview/redhatopenshift.json:
```diff
@@ -0,0 +1,2732 @@
+        }
+      }
+    },
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.RedHatOpenShift/openShiftCluster/{resourceName}/machinePools": {
```
>openShiftCluster [](http://example.com/codeflow?start=108&length=16)

[ARMBlockingComment] openShiftClusters (with an 's')

cadenmarchese replied:
see comment https://github.com/Azure/azure-rest-api-specs/pull/30004#discussion_r1887283477

mentat9 commented on specification/redhatopenshift/resource-manager/Microsoft.RedHatOpenShift/openshiftclusters/preview/2024-08-12-preview/redhatopenshift.json:
```diff
@@ -0,0 +1,2732 @@
+        }
+      }
+    },
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.RedHatOpenShift/openShiftCluster/{resourceName}/syncSets": {
```
>openShiftCluster [](http://example.com/codeflow?start=108&length=16)

[ARMBlockingComment] Name must be plural and match collection GET (openShiftClusters).

cadenmarchese replied:
see comment https://github.com/Azure/azure-rest-api-specs/pull/30004#discussion_r1887283477

mentat9 commented on specification/redhatopenshift/resource-manager/Microsoft.RedHatOpenShift/openshiftclusters/preview/2024-08-12-preview/redhatopenshift.json:
```diff
@@ -0,0 +1,2732 @@
+        }
+      }
+    },
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.RedHatOpenShift/openshiftclusters/{resourceName}/syncIdentityProvider/{childResourceName}": {
```
>syncIdentityProvider [](http://example.com/codeflow?start=141&length=20)

[ARMBlockingComment] syncIdentityProviders

cadenmarchese replied:
see comment https://github.com/Azure/azure-rest-api-specs/pull/30004#discussion_r1887283477

mentat9 commented on specification/redhatopenshift/resource-manager/Microsoft.RedHatOpenShift/openshiftclusters/preview/2024-08-12-preview/redhatopenshift.json:
```diff
@@ -0,0 +1,2732 @@
+        }
+      }
+    },
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.RedHatOpenShift/openshiftclusters/{resourceName}/syncSet/{childResourceName}": {
```
>syncSet [](http://example.com/codeflow?start=141&length=7)

[ARMBlockingComment] syncSets

mentat9 commented:
@cadenmarchese - Earlier commits in this PR were signed off by ARM, but currently none of those commits are in the PR (it only has 9 commits, all later and showing all newly added files). There is no way to tell what was reviewed and what wasn't. Also, the APIs currently in the PR cannot be signed off: there are many blocking issues that need to be addressed. I've left review comments.

cadenmarchese committed:
add where clause to suppressions, add reason

cadenmarchese committed:
fix CI suppressions (where and from clauses)

cadenmarchese committed:
Resolve ARM feedback
* add singleton GETs for openShiftVersions and PWIRS
* remove unused resource types machinePool, syncSet, secret, and
syncIdentityProvider after a conversation with my team, as they
were never used, there are no routes in our API to handle them,
and they are not included in our ARM manifests
* camelCase fixes

cadenmarchese committed:
resolve Swagger ModelValidation

rkmanda reviewed (approved):
:shipit:

cadenmarchese committed:
convert cspell to new format (see PR#32171)

cadenmarchese committed:
add 'oidc' to cspell to resolve swagger spell check CI failure
